apps/confidential/confidential_test.go

```diff
@@ -529,6 +531,185 @@ func TestNewCredFromTokenProviderError(t *testing.T) {
+			}
+			var actual map[string]any
+			if err := json.Unmarshal([]byte(claims[0]), &actual); err != nil {
+				t.Fatal(err)
+			}
+			if diff := pretty.Compare(expected, actual); diff != "" {
+				t.Fatal(diff)
+			}
+		}
+		for _, method := range []string{"authcode", "authcodeURL", "credential", "obo"} {
+			t.Run(method, func(t *testing.T) {
+				mockClient := mock.Client{}
+				clientInfo, idToken, refreshToken := "", "", ""
+				if method == "obo" {
+					clientInfo = base64.RawStdEncoding.EncodeToString([]byte(`{"uid":"uid","utid":"utid"}`))
+					idToken = mock.GetIDToken(tenant, authority)
+					refreshToken = "rt"
+					// TODO: OBO does instance discovery twice before first token request https://github.com/AzureAD/microsoft-authentication-library-for-go/issues/351
+					mockClient.AppendResponse(mock.WithBody(mock.GetInstanceDiscoveryBody(lmo, tenant)))
+				}
+				mockClient.AppendResponse(mock.WithBody(mock.GetInstanceDiscoveryBody(lmo, tenant)))
+				mockClient.AppendResponse(mock.WithBody(mock.GetTenantDiscoveryBody(lmo, tenant)))
+				mockClient.AppendResponse(
+					mock.WithBody(mock.GetAccessTokenBody(accessToken, idToken, refreshToken, clientInfo, 3600)),
+					mock.WithCallback(func(r *http.Request) {
+						if err := r.ParseForm(); err != nil {
+							t.Fatal(err)
+						}
+						validate(t, r.Form)
+					}),
+				)
+				client, err := New("client-id", cred, WithAuthority(authority), WithClientCapabilities(test.capabilities), WithHTTPClient(&mockClient))
+				if err != nil {
+					t.Fatal(err)
+				}
+				if _, err = client.AcquireTokenSilent(context.Background(), tokenScope); err == nil {
+					t.Fatal("silent authentication should fail because the cache is empty")
+				}
+				ctx := context.Background()
+				var ar AuthResult
+				switch method {
+				case "authcode":
+					ar, err = client.AcquireTokenByAuthCode(ctx, "code", "https://localhost", tokenScope, WithClaims(test.claims))
+				case "authcodeURL":
+					u := ""
+					if u, err = client.AuthCodeURL(ctx, "client-id", "https://localhost", tokenScope, WithClaims(test.claims)); err == nil {
+						var parsed *url.URL
+						if parsed, err = url.Parse(u); err == nil {
+							validate(t, parsed.Query())
+							return // didn't acquire a token, no need for further validation
+						}
+					}
+				case "credential":
+					ar, err = client.AcquireTokenByCredential(ctx, tokenScope, WithClaims(test.claims))
+				case "obo":
+					ar, err = client.AcquireTokenOnBehalfOf(ctx, "assertion", tokenScope, WithClaims(test.claims))
+				default:
+					t.Fatalf("test bug: no test for " + method)
+				}
+				if err != nil {
+					t.Fatal(err)
+				}
+				if ar.AccessToken != accessToken {
+					t.Fatalf(`unexpected access token "%s"`, ar.AccessToken)
+				}
+				// silent auth should now succeed, provided no claims are requested, because the client has cached an access token
+				if method == "obo" {
+					ar, err = client.AcquireTokenOnBehalfOf(ctx, "assertion", tokenScope)
```

Comment:
This is to verify the client has a cached access token, because below we want to validate that `AcquireTokenOnBehalfOf` doesn't return this token when called with claims. (`AcquireTokenOnBehalfOf` handles both silent and non-silent auth; `AcquireTokenSilent` accesses a different cache and doesn't take a user assertion argument.)